### PR TITLE
[WIP] [checkpoint v2] Use remote uploader v2 for checkpointing

### PR DESCRIPTION
**Pull request by bigning**
# what?

Composer change to use the new RemoteUploader for saving checkpoing

# Test
llm-foundry PR: https://github.com/mosaicml/llm-foundry/pull/1237
## 1-node small model OCI test
test-uploader-wfA8Vz
## 1-node small model S3 test
save: test-uploader-10udLt e2e time: 3 min 40s
resume: test-uploader-TDGgTs

use old remote_uploader_downloader `test-uploader-racGuC` e2e time: 3 min 42s

### CODE REVIEW

File: composer/checkpoint/state_dict.py [hunk 1]
```diff
@@ -152,3 +154,94 @@ def _get_model_state_dict_with_fsdp_context_manager(model: nn.Module, sharded_st
...
         model_state_dict = model.state_dict()
     return model_state_dict
+
+
+def get_metadata_state_dict(
+    model: Optional[Union[ComposerModel, nn.Module]] = None,
+    sharded_state_dict: Optional[bool] = None,
+    precision: Optional[Union[str, torch.dtype]] = None,
+    device: Optional[Device] = None,
+    device_train_microbatch_size: Optional[int] = None,
```
Comment (mvpatel2000) [inline composer/checkpoint/state_dict.py:164]: Is this copied from Evan's PR?

**Review by mvpatel2000 — COMMENTED** [review]
Why do we default to v1 still?

File: composer/checkpoint/state_dict.py (hunk 1 above)
Comment (bigning) [reply composer/checkpoint/state_dict.py:164]: oh it got messed up. will clean up

File: composer/callbacks/checkpoint_saver.py [hunk 2]
```diff
@@ -520,11 +549,18 @@ def _save_checkpoint(self, state: State, logger: Logger):
...
-                        logger.upload_file(
-                            remote_file_name=symlink_name,
-                            file_path=symlink_filename,
-                            overwrite=True,
-                        )
+                        if self.use_remote_uploader_v2 and self.remote_uploader is not None:
+                            self.remote_uploader.upload_file_async(
+                                remote_file_name=symlink_name,
+                                file_path=symlink_filename,
+                                overwrite=True,
```
Comment (eracah) [inline composer/callbacks/checkpoint_saver.py:556]: ok this makes sense for a PR, where we just simplify the remoteuploader and don't change any functionality, but eventually we want to have the remoteuploader manage symlinks. That way we can actually avoid race conditions

**Comment by bigning** [comment]
Discussed with @eracah , we're going to add a new checkpoint saver, instead of changing current one.

**Pull request closed without merging**